project/src/main/java/com/google/sps/servlets/BuyServlet.java

```diff
@@ -0,0 +1,128 @@
+        DataSource pool = (DataSource) request.getServletContext().getAttribute("db-connection-pool");
+
+        try (Connection conn = pool.getConnection()) {
+            long user = Long.parseLong(request.getParameter("user"));
+            long competition = Long.parseLong(request.getParameter("competition"));
+            String searchQuery = request.getParameter("search");
+            int amtInvested = Integer.parseInt(request.getParameter("amount"));
+            Investment investment = addInvestment(conn, user, competition, searchQuery, amtInvested);
+            if (investment == null) {
+                response.setStatus(500);
+                response.getWriter().write("Timeout while fetching investment. Try again later.");
+            } else {
+                Gson gson = new Gson();
+                response.setContentType("application/json");
+                response.getWriter().println(gson.toJson(investment));
+            }
+        } catch (SQLException ex) {
+            LOGGER.log(Level.WARNING, "Error while attempting to buy investment.", ex);
+            response.setStatus(500);
+            response.getWriter().write("Unable to successfully add investment");
+        }
+    }
+
+    private Investment addInvestment (
+        Connection conn, long user, long competition, String searchQuery, int amtInvested
+        ) throws SQLException, IOException {
+        
+        InvestmentCalculator calc = new InvestmentCalculator();
+        Long currentDateSeconds = calc.getLatestDate();
+        Date currentDate = new Date(currentDateSeconds * 1000);
+
+        // check if we already have data for this search
+        ImmutableList<Long> data = calc.getInvestmentDataIfExists(searchQuery);
+        if (data.isEmpty()) {
+            // if not, fetch using cloud function and wait for some set time
+            data = calc.createAndFetchInvestmentData(searchQuery, data);
+        }
+
+        if (data.isEmpty()) {
+            //we timed out. return null and then send response - DO NOT add to investments db
+            return null;
+        }
+
+        String stmt = String.format("INSERT INTO investments (user, competition, google_search, " 
+            + "invest_date, sell_date, amt_invested) VALUES (%d, %d, '%s', DATE '%tF', NULL, %d);",
+            user, competition, searchQuery, currentDate, amtInvested);
+
+        try (PreparedStatement investmentStmt = conn.prepareStatement(stmt);) {
+            // Execute the statement
+            investmentStmt.execute();
+        }
+        String getId = "SELECT LAST_INSERT_ID();";
+        try (PreparedStatement idStmt = conn.prepareStatement(getId);) {
+            ResultSet rs = idStmt.executeQuery();
+            long id;
+            while (rs.next()) {
+                id = rs.getInt(1);
+                LOGGER.log(Level.INFO, "Investment " + id + " added to database.");
+                updateAmountAvailable(conn, user, competition, amtInvested);
```

Comment:
If this step fails, would the user have had a free purchase OR is there some other reconciliation mechanism?

Otherwise, SQL transactions

Reply:
Yeah they will get a free purchase. What do I do about that?

Reply:
SQL transactions. This has a good overview - https://www.mysqltutorial.org/mysql-jdbc-transaction/

Reply:
Done